internal/buildjob/buildjob.go

```diff
@@ -187,78 +187,155 @@ func (j *Job) generateBuildOptions(ctx context.Context, cib *v1alpha1.ContainerI
+					NonSSL:   false,
+					Username: authConfig.Username,
+					Password: authConfig.Password,
+				}
+			}
+
 		case apiReg.DynamicCloudCredentials:
-			fetchAuth = func() (string, string, error) {
-				return j.getDockerAuthFromFS(apiReg.Server)
+			authConfigs, err := j.getDockerAuthsFromFS()
+			if err != nil {
+				return nil, err
 			}
-		}
 
-		if fetchAuth != nil {
-			var err error
-			conf.Username, conf.Password, err = fetchAuth()
+			username, password, err := j.getBasicAuthForHost(authConfigs, apiReg.Server)
 			if err != nil {
 				return nil, err
 			}
+
+			logNewRegistryConfig(apiReg.Server, "dynamic cloud credentials")
+			configuredRegistries[apiReg.Server] = config.Registry{
+				Host:     apiReg.Server,
+				NonSSL:   apiReg.NonSSL,
+				Username: username,
+				Password: password,
+			}
+
+		default:
+			// If no recognizable auth config is present, configure registry without authentication.
+			logNewRegistryConfig(apiReg.Server, "no source, configured without auth")
+			configuredRegistries[apiReg.Server] = config.Registry{
+				Host:   apiReg.Server,
+				NonSSL: apiReg.NonSSL,
+			}
 		}
+	}
 
-		configs = append(configs, conf)
+	registryConfigs = []config.Registry{}
+	for _, v := range configuredRegistries {
+		registryConfigs = append(registryConfigs, v)
 	}
 
-	return configs, nil
+	return registryConfigs, nil
 }
 
-func (j *Job) getDockerAuthFromFS(host string) (string, string, error) {
+func (j *Job) getDockerAuthsFromFS() (credentials.AuthConfigs, error) {
 	if _, err := os.Stat(config.DynamicCredentialsFilepath); os.IsNotExist(err) {
-		return "", "", errors.Wrap(err, "filesystem docker credentials missing")
+		return nil, errors.Wrap(err, "filesystem docker credentials missing")
 	}
 
 	input, err := ioutil.ReadFile(config.DynamicCredentialsFilepath)
 	if err != nil {
-		return "", "", errors.Wrap(err, "cannot read docker credentials")
+		return nil, errors.Wrap(err, "cannot read docker credentials")
 	}
 
-	username, password, err := credentials.ExtractDockerAuth(input, host)
+	authConfigs, err := credentials.ExtractAuthConfigs(input)
 	if err != nil {
-		return "", "", errors.Wrapf(err, "cannot process filesystem docker auth for host %q", host)
+		return nil, errors.Wrapf(err, "Failed to read docker auth from filesystem")
 	}
 
-	return username, password, nil
+	return authConfigs, nil
 }
 
-func (j *Job) getDockerAuthFromSecret(ctx context.Context, host, name, namespace string) (string, string, error) {
-	secret, err := j.clientk8s.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
+func (j *Job) getDockerAuthsFromSecret(ctx context.Context, secretName string, secretNamespace string) (credentials.AuthConfigs, error) {
+	secret, err := j.clientk8s.CoreV1().Secrets(secretNamespace).Get(ctx, secretName, metav1.GetOptions{})
 	if err != nil {
-		return "", "", errors.Wrap(err, "cannot find registry auth secret")
+		return nil, errors.Wrap(err, "cannot find registry auth secret")
```

Comment:
FYI this was the original error message (I only updated the first part of the tuple). I have updated it because you're right, but generally I try to minimize changes to existing code as part of any specific PR

Reply:
Sure. But I find its also the only time you recognize past mistakes. And given it was touched it becomes fair game.